flink-core/src/test/java/org/apache/flink/api/common/typeutils/TypeSerializerUpgradeTestBase.java

```diff
@@ -46,28 +50,25 @@
 /**
  * A test base for testing {@link TypeSerializer} upgrades.
  *
- * <p>You can run {@link #generateTestSetupFiles()} on a Flink branch to (re-)generate the test data
- * files.
+ * <p>You can run {@link
+ * #generateTestSetupFiles(org.apache.flink.api.common.typeutils.TypeSerializerUpgradeTestBase.TestSpecification)}
+ * on a Flink branch to (re-)generate the test data files.
  */
-public abstract class TypeSerializerUpgradeTestBase<PreviousElementT, UpgradedElementT>
-        extends TestLogger {
+@ExtendWith(TestLoggerExtension.class)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
```

Comment:
Interesting: I didn't realize this lets you use abstract `testData()` as a `MethodSource`, which is perfect!

Reply:
Thanks @zentol for help with finding such solution at https://github.com/apache/flink/pull/19753#discussion_r876860090